cloud/ovirt: set storage domain when creating vm

tareqalayan opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request
 - New Module Pull Request
 - Bugfix Pull Request
 - Docs Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```

```

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

<!-- Paste verbatim command output below, e.g. before and after your change -->
```

```

tareqalayan commented:
@machacekondra please review thanks

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -444,8 +447,36 @@ def __get_template_with_version(self):
 
         return template
 
+    def __get_storage_domain(self):
```
You may use `search_by_name()' method:

```
search_by_name(
    self._connection.system_service().storage_domains_service(),
    self._module.params['storage_domain']
)
```

https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/ovirt.py#L221

tareqalayan replied:
ok will change this

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -444,8 +447,36 @@ def __get_template_with_version(self):
+        disk_attachments = self._connection.follow_link(
+            template_service.get().disk_attachments
+        )
+        disk = disk_attachments[0].disk
```
Hmm, what if the template will have more disks. We have to somehow distinguish which one should go where...

tareqalayan replied:
I think they all should go there, no?

tareqalayan replied:
done

tareqalayan commented:
@machacekondra please review 

thanks,

machacekondra commented:
We need to think about the possibility of specification of specific storage of specific disk.

machacekondra commented:
Maybe we can do following:

Specific disk to specific SD:
```yaml
ovirt_vms:
  name: myvm
  template: mytemplate
  cluster: mycluster
  disks:
    -  id: 123
       storage_domain: nfs_0
```

All disks to one SD:
```yaml
ovirt_vms:
  name: myvm
  template: mytemplate
  cluster: mycluster
  disks:
    - storage_domain: nfs_0
```
What do you think?

tareqalayan commented:
@machacekondra 
OK sounds good, 
I will do the change during the week or beginning of next week.

mattclay commented:
Closing and re-opening to trigger CI.

jctanner commented:
bot_status

tareqalayan commented:
@machacekondra 

i think if there is template in the module we don't have to specify disk_id or name we just need to copy all disks attached to template to the specified storage domain? 

i am just thinking loud here ... what do you think?

tareqalayan committed:
cloud/ovirt: set storage domain when creating vm

Now you can specify the storage where to store the vm disks.

For example if you want to create a VM from template into specific
storage domain you can do the following:

  ovirt_vms:
    name: vm_on_my_storage_domain
    cluster: my_cluster
    template: my_template
    operating_system: other_linux
    type: server
    cpu_cores: 1
    cpu_sockets: 1
    state: stopped
    clone: True
    disks:
      - id: 1111
        storage_domain: my_storage_domain

tareqalayan commented:
@machacekondra 
Ondra can you please review?

myakove commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -477,8 +478,55 @@ def __get_template_with_version(self):
 
+    def __get_storage_domain(self, disk_id):
+        storage_domain = None
+
```
disks = self._module.params.get('disks', list())

tareqalayan replied:
done

myakove commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -477,8 +478,55 @@ def __get_template_with_version(self):
+        storage_domain = None
+
+        if self._module.params['disks']:
+            for disk in self._module.params['disks']:
```
for disk in disks:

tareqalayan replied:
done

myakove commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -477,8 +478,55 @@ def __get_template_with_version(self):
+
+    def __get_disk_attachments(self, template):
+
+        # if self.__get_storage_domain() is None:
```
remove comments please

tareqalayan replied:
done

myakove commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -539,6 +587,7 @@ def build_entity(self):
                 self.param('serial_policy') is not None or
                 self.param('serial_policy_value') is not None
             ) else None,
+            disk_attachments=disks if disks else None,
```
disks or None

tareqalayan replied:
I will leave this as is to be like the lines above.

machacekondra replied:
I wonder what should happen when disks is [].

tareqalayan replied:
nothing should happen it put the disk on the first storage storage domain on the list ... in my case it always choose nfs_0 ....

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -477,8 +478,52 @@ def __get_template_with_version(self):
+
+        disk_attachments = self._connection.follow_link(
+            template_obj.get().disk_attachments
+        )
```
No need to re-fetch the template, you can just write:
```python
disk_attachments = self._connection.follow_link(template.disk_attachments)
```

tareqalayan replied:
ack

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -477,8 +478,52 @@ def __get_template_with_version(self):
+                otypes.DiskAttachment(
+                    disk=otypes.Disk(
+                        id=disk.id,
+                        format=otypes.DiskFormat.COW,
```
What about 'raw'?

tareqalayan replied:
I didn't put much thinking here ... what it should it be?

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -477,8 +478,52 @@ def __get_template_with_version(self):
+
+        for disk in disks:
+            if disk_id == disk.get('id'):
+                storage_domain = search_by_name(
```
Better to use ```get_id_by_name``` it get proper error message if storage is not found

tareqalayan replied:
ack

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -709,6 +755,9 @@ def wait_for_down(self, vm):
     def __attach_disks(self, entity):
         disks_service = self._connection.system_service().disks_service()
 
+        if self.param('template'):
```
What if we want to create VM from template, and attach some other disks?

tareqalayan replied:
i am not familiar with this use case. 

do we want to support that ? 

should i  grab all template disks + disks stated in the module?

machacekondra commented on lib/ansible/modules/cloud/ovirt/ovirt_vms.py:
```diff
@@ -477,8 +478,52 @@ def __get_template_with_version(self):
+        disks = self._module.params.get('disks', list())
+
+        for disk in disks:
+            if disk_id == disk.get('id'):
```
Can we somehow implement behavior of specify storage_domain for all disks? If I understand the code correctly, it not currently.

tareqalayan replied:
The idea here is to specify which disk go where see example:
  ovirt_vms:
    name: vm_test
    cluster: clu1
    template: template01
    operating_system: other_linux
    type: server
    cpu_cores: 1
    cpu_sockets: 1
    state: stopped
    clone: True
    timeout: 600
    disks:
      - id: disk_id1
        storage_domain: my_nfs_1
     -  id: disk_id1
        storage_domains: my_nfs_3

i would prefer not to decalre any disk and only storage domain so all disks related to template will go to spesfic storage domain ... 
but you said there is a use case to split disks of templates on multiple storage domains .... 

please tell me what you suggest... and i will implement it ...

tareqalayan replied:
the plan now is: 
1. if there is a template specified and a storage domain: all disks related to that template will go into the storage domain
2. if there is a template specifeid and disks is specified each disk will go to the specified storage domain:
template: my_template
disks:
  - id_of_disk
  - storage domain 


@machacekondra please ack

machacekondra reviewed (changes requested):
Can you also verify the beahvior of copy/move disks?
